README.md

```diff
@@ -1,12 +1,201 @@
+    endpoint.Decode,
+    endpoint.Encode,
+    endpoint.EncodeError,
+    ep,
+  )
+
+  lambda.StartHandlerFunc(h)
+}
+```
+
+The above example is a lambda that is triggered by an API Gateway request. The layers
+could look like this:
+
+| Layer | Model |
+|-------|-------|
+| Handler | API Gateway Request & Response |
+| Endpoint | JSON |
+| Service | internal model |
+
+Another example could be a lambda triggered by a DynamoDB Stream
+
+ Layer | Model |
+|-------|--------|
+| Handler | DynamoDB Stream Record |
+| Endpoint | database model |
+| Service | internal model |
+
+or an HTTP Server with a Protobuf body
+
+| Layer | Model |
+|-------|--------|
+| Handler | HTTP Request & Response |
+| Endpoint | protobuf model |
+| Service | internal model |
+
+## Usage for clients
+
+For clients the same principle applies as above but from the rule
+that dependencies only point inwards, we can conclude that instead of wrapping
+the inner layers with the next outer layer, we provide clients that can deal
+with the internal model. That leads to thin wrappers around provided libraries
+like the aws-sdk or HTTP clients. The layers in this case look a bit different
+and I inverted the order to better represent the flow of data through them.
+
+| Layer | Model |
+|-------|-------|
+| Service | internal model | businesss logic |
+| Endpoint | DTO | DTO encoding, meta-data injection |
+| Handler | transport protocol | retries, backoffs, meta-data injection, error handling |
+
+An example for an aws-sdk implementation could look like this
+
+```go
+import (
+...
+  "github.com/aws/aws-sdk-go-v2/service/dynamodb"
+  "github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
+...
+)
+  type dynamoDBClient interface {
+    GetItem(ctx context.Context, params *dynamodb.GetItemInput, optFns ...func(*dynamodb.Options)) (*dynamodb.GetItemOutput, error)
+  }
+
+  type db struct {
+    client dynamoDBClient
+  }
+
+  func New() *db {
+    cfg, err := config.LoadDefaultConfig(context.TODO(), config.WithRegion("us-west-2"))
+    if err != nil {
+        log.Fatalf("unable to load SDK config, %v", err)
+    }
+
+    client := dynamodb.NewFromConfig(cfg)
+    return &db{
+      client: client, 
+    }
+  }
+
+  func (d *db) GetCustomer(ctx context.Context, custmerID string) (service.Customer, error) {
+    input := dynamodb.GetItemInput{
+      TableName: "My-Table",
+      Key: map[string]types.AttributeValue{
+       "id": customerID,
+      }
+    }
+
+    out, err := d.client.GetItem(ctx, input)
+    if err != nil {
+      return service.Customer{}, err
+    }
+
+    return dbModelToInternalModel(out)
+  }
+```
+
+In most SDK examples you will not have to deal with the transport protocol and it
+is hidden inside of the SDK. In other cases you will need to provide that yourself
+and it makes sense to separate it in the same way. Imagine an HTTP service provided
+by, the HTTP layer with retries, jitter, exponential back-offs should be separated
```

Comment:
Provided by who? :suspense:

Reply:
Hmm I don't remember anymore 😅 I guess the `provided by` part can go out?